src/CBF/CBF.jl

```diff
@@ -0,0 +1,588 @@
+            end
+            @assert k == length(vars)
+        end
+
+        for ci in getmodelcons(MOI.VectorAffineFunction{Float64},
+            MOI.PositiveSemidefiniteConeTriangle) # MOI affine constraints.
+            side = getconset(ci).side_dimension
+            push!(psdcon, side)
+            fi = getconfun(ci)
+            for vt in fi.terms
+                st = vt.scalar_term
+                # Get (i,j) index in symmetric matrix from lower triangle index.
+                i = div(1 + isqrt(8*vt.output_index - 7), 2)
+                j = vt.output_index - div(i*(i-1), 2)
+                push!(hcoord, (length(psdcon), st.variable_index.value, i, j,
+                    st.coefficient))
+            end
+            k = 0
+            for i in 1:side, j in 1:i
+                k += 1
+                if !iszero(fi.constants[k])
+                    push!(dcoord, (length(psdcon), i, j, fi.constants[k]))
+                end
+            end
+            @assert k == MOI.output_dimension(fi)
+        end
+
+        if !isempty(psdcon)
+            println(io, "PSDCON")
+            println(io, length(psdcon))
+            for v in psdcon
+                println(io, v)
+            end
+            println(io)
+        end
+
+        if !isempty(hcoord)
+            println(io, "HCOORD")
+            println(io, length(hcoord))
+            for (a, b, c, d, v) in hcoord
+                println(io, a-1, " ", b-1, " ", c-1, " ", d-1, " ", v)
+            end
+            println(io)
+        end
+
+        if !isempty(dcoord)
+            println(io, "DCOORD")
+            println(io, length(dcoord))
+            for (a, b, c, v) in dcoord
+                println(io, a-1, " ", b-1, " ", c-1, " ", v)
+            end
+            println(io)
+        end
+    end
+    
+    return
+end
+
+# ==============================================================================
+#
+#   MOI.read_from_file
+#
+# The CBF file format (version 3) is described at
+# http://cblib.zib.de/doc/format3.pdf.
+#
+# ==============================================================================
+
+function mattovecidx(i, j)
+    if i < j
+        (i,j) = (j,i)
+    end
+    return div((i-1)*i, 2) + j
+end
+
+function MOI.read_from_file(model::Model, filename::String)
+    if !MOI.is_empty(model)
+        error("Cannot read in file because model is not empty.")
+    end
+
+    x = MOI.VariableIndex[]
+    X = Vector{MOI.VariableIndex}[]
+    objterms = MOI.ScalarAffineTerm{Float64}[]
+    objconst = 0.0
+    conecons = Tuple{String, Int}[]
+    conterms = Vector{MOI.ScalarAffineTerm{Float64}}[]
+    conconst = Float64[]
+    psdcons = Tuple{Int, Int}[]
+    psdterms = Vector{MOI.ScalarAffineTerm{Float64}}[]
+    psdconst = Float64[]
+
+    open(filename, "r") do io
+        while !eof(io)
+            line = strip(readline(io))
+
+            # Comments.
+            if startswith(line, "#") || length(line) == 1
```

Comment:
Won't this skip lines with a single character like `1`, or a name like `x`? `length(line) == 0`?

Reply:
we won't see such lines
https://github.com/JuliaOpt/ConicBenchmarkUtilities.jl/blob/4267bfedeace99fcc4a03256eef8d44151f38b95/src/cbf_input.jl#L47

Reply:
ConicBenchmarkUtilities doesn't `strip` the line. So an empty line will be `"\n"` (or potentially even `"\r\n"`), thus the length one (or two). An empty line here will just be `""`.

Reply:
True that